File: csharp/NativeUtils/FileJanitor.cs
#if NETSTANDARD1_0 || NETSTANDARD1_1 || NETSTANDARD1_2 || NETSTANDARD1_3 || NETSTANDARD1_4 || NETSTANDARD1_5 || NETSTANDARD1_6 || NETCOREAPP1_0 || NETCOREAPP1_1
#define NETSTANDARD_BELOW_2_0
#endif

using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

namespace RTMath.Utilities
{
	using static ResourceLoaderUtils.Util;
	using FileStream = ResourceLoaderUtils.FileStream;

	internal class FileJanitor
	{
		private const string LockFileName = "lockfile.$$$";
		private static readonly object CleanupLock = new object();
		private static readonly List<CleanupPath> CleanupDirs = new List<CleanupPath>();

		internal bool IsIgnoredException(Exception e)
		{
			return e is UnauthorizedAccessException || e is System.Security.SecurityException || e is IOException;
		}

		private static bool IsLockFile(string path)
		{
			return path.EndsWith(LockFileName);
		}

		public static FileStream TryOpenWriteable(string path) => TryOpenForWriteTest(path);

		public static string LockFilePath(string dir) => Path.Combine(dir, LockFileName);

		public static FileStream TryCreateLockFile(string dir)
		{
			return TryOpenForWriteTest(LockFilePath(dir), FileOptions.DeleteOnClose);
		}

		public static bool LockFileExists(string dir) => File.Exists(LockFilePath(dir));

		public static DateTime LockFileWriteTime(string dir)
		{
			try
			{
				return File.GetLastWriteTimeUtc(LockFilePath(dir));
			}
			catch
			{
				return DateTime.MinValue;
			}
		}

		/// Delete directory carefully, only if _none_ of the files in it are opened by someone else
		/// If at least one file is locked, the operation silently fails without deleting _anything_
		/// It is safe to call this operation concurrently on a single directory, but if the directory contents are modified
		/// by the code that does not respect our lock file, its contents may still be deleted partially and false will be returned
		public static bool TryDeleteDirectory(string dir)
		{
			bool isSuccess = false;
			using (var lockFile = TryCreateLockFile(dir))
			{
				if (null == lockFile)
					return false;

				var files = Directory.EnumerateFiles(dir);
				var opened = new List<FileStream>();

				foreach (string fname in files)
					if (!IsLockFile(fname))
					{
						var f = TryOpenWriteable(fname);
						if (null == f)
							goto Fail;

						opened.Add(f);
					}

				isSuccess = true;

				Fail:
				foreach (var f in opened)
				{
					f.Dispose();
				}

				if (isSuccess)
				{
					foreach (string fname in files)
					{
						if (!IsLockFile(fname) && !TryDeleteFile(fname))
							return false;
					}
				}
			}

			if (isSuccess)
			{
				try
				{
					Directory.Delete(dir);
					return true;
				}
				catch { }
			}

			return false;
		}


		public static void TryCleanup(string dir, bool cleanDir = true, string subDirRegEx = null)
		{
			new CleanupPath(dir, cleanDir, subDirRegEx).TryCleanup();
		}


		public static void TryCleanup()
		{
			lock (CleanupLock)
			{
				var deleted = new List<CleanupPath>();
				foreach (CleanupPath p in CleanupDirs)
					if (p.TryCleanup())
						deleted.Add(p);

				foreach (CleanupPath p in deleted)
					CleanupDirs.Remove(p);
			}
		}

		/// <summary>
		/// Register path for cleanup.
		/// </summary>
		/// <param name="path">path to clean</param>
		/// <param name="cleanDir">will clean the specified directory, ignoring subdirectories (default)</param>
		/// <param name="subdirRegEx">will apply the same cleanup logic to all subdirectories found in the path (not-recursive), but not the path itself</param>
		public static void AddCleanupPath(string path, bool cleanDir = true, string subdirRegEx = null)
		{
			lock (CleanupLock)
			{
				CleanupDirs.Add(new CleanupPath(path, cleanDir, subdirRegEx));
			}
		}

#if !NETSTANDARD_BELOW_2_0
		/// <summary>
		/// Register cleanup callback
		/// Cleanup will be called automatically on events DomainUnload and ProcessExit, unless netstandard &lt; 2.0 or .net core &lt; 2.0
		/// each added path is only cleaned if none of the files in it are open or locked
		/// </summary>
		public static void RegisterForCleanupOnExit()
		{
			lock (CleanupLock)
			{
				if (!_handlerRegistered)
				{
					AppDomain.CurrentDomain.DomainUnload += (sender, eventArgs) => TryCleanup();
					AppDomain.CurrentDomain.ProcessExit += (sender, eventArgs) => TryCleanup();
					_handlerRegistered = true;

				}
			}
		}

		private static bool _handlerRegistered;
#endif // !NETSTANDARD_BELOW_2_0
	}

	internal class CleanupPath
	{
		private const int CleanDir = 1;

		private readonly String _path;
		private readonly String _subDirRegEx;
		private readonly int _flags;

		public bool TryCleanup()
		{
			try
			{
				if (!Directory.Exists(_path))
					return true;

				bool success = true;
				// Clean subdirs?
				if (null != _subDirRegEx)
				{
					var dirs = Directory.EnumerateDirectories(_path);
					Regex regex = new Regex(_subDirRegEx);
					foreach (var dir in dirs)
						if (regex.IsMatch(Path.GetFileNameWithoutExtension(dir)))
							success &= FileJanitor.TryDeleteDirectory(dir);
				}

				if (0 != (_flags & CleanDir))
					success &= FileJanitor.TryDeleteDirectory(_path);

				return success;
			}
			catch
			{
				return false;
			}
		}

		public CleanupPath(string path, bool cleanDir, string subDirRegEx)
		{
			_path = path;
			_subDirRegEx = subDirRegEx;
			_flags = (cleanDir ? CleanDir : 0);
		}
	}
}

File: java/main/src/main/java/rtmath/utilities/FileJanitor.java
package rtmath.utilities;

import java.io.Closeable;
import java.io.IOException;
import java.nio.channels.FileChannel;
import java.nio.channels.FileLock;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static rtmath.utilities.ResourceLoaderUtils.closeLockedFileChannel;
import static rtmath.utilities.ResourceLoaderUtils.openLockedFileChannel;

/**
 * Set of helper methods mostly for checking directory access locks and performing cleanup.
 */
public class FileJanitor {
    private static final String lockFileName = "lockfile.$$$";

    private static boolean isLockFile(Path path) {
        return path.getFileName().endsWith(lockFileName);
    }

    private static FileLock tryOpenForWriteTest(Path path) {

        // NOTE: We previously had DELETE_ON_CLOSE here but on Linux the file is deleted immediately after opening!
        // Delete the file yourself, please
        try {
            return openLockedFileChannel(path
                , StandardOpenOption.CREATE
                , StandardOpenOption.READ
                , StandardOpenOption.WRITE
            );
        } catch (IOException e) {
            return null;
        }
    }

    static boolean tryDelete(Path path) {
        try {
            Files.delete(path);
            return true;
        }
        catch (IOException e) {
            return false;
        }
    }


    public static Path lockFilePath(Path dir) {
        return dir.resolve(lockFileName);
    }

    static class LockFile implements Closeable {
        Path path;
        FileLock lock;

        LockFile(FileLock lock, Path path) {
            this.lock = lock;
            this.path = path;
        }


        @Override
        public void close() {

            if (null != lock) {
                closeLockedFileChannel(lock);
                tryDelete(path);
                lock = null;
            }
        }

        protected void finalize() throws Throwable {
            close();
        }

        public FileChannel channel() {
            return lock.channel();
        }
    }

    public static LockFile tryCreateLockFile(Path dir) {
        dir = lockFilePath(dir);
        FileLock lock = tryOpenForWriteTest(dir);
        return null != lock ? new LockFile(lock, dir) : null;
    }

    public static boolean lockFileExists(Path dir)
    {
        return Files.exists(lockFilePath(dir));
    }

    public static long getLockFileWriteTime(Path path) {
        try {
            BasicFileAttributes attr = Files.readAttributes(path, BasicFileAttributes.class);
            return attr.lastAccessTime().toMillis();
        }
        catch (IOException e) {
            return Long.MIN_VALUE;
        }
    }

    /**
     * Delete directory carefully, only if _none_ of the files in it are opened by someone else.
     * <p>
     * If at least one file is locked, the operation silently fails without deleting _anything_.
     * <p>
     * It is safe to call this operation concurrently on a single directory, but if the directory contents are modified
     * by the code that does not respect our lock file, its contents may still be deleted partially
     * and false will be returned.
     *
     * @param dir path being deleted
     * @return <code>true</code> if was able to delete all the files in the directory without triggering any locks.
     * <code>false</code> in no changes were made to the contents of the directory (except the case described above).
     */
    public static boolean tryDeleteDirectory(Path dir) {
        boolean isSuccess = false;
        LockFile lock = tryCreateLockFile(dir);

        if (null == lock)
            return false;

        try {
            List<FileLock> openedFiles = new ArrayList<>();
            ArrayList<Path> foundFiles = new ArrayList<>();

            ListDeleteableFiles:
            try (DirectoryStream<Path> paths = Files.newDirectoryStream(dir)) {
                for (Path path : paths) {
                    if (isLockFile(path))
                        continue;

                    FileLock fl;
                    if (Files.isDirectory(path) || null == (fl = tryOpenForWriteTest(path)))
                        break ListDeleteableFiles;

                    openedFiles.add(fl);
                    foundFiles.add(path);
                }

                isSuccess = true;
            }

            for (FileLock fl : openedFiles)
                closeLockedFileChannel(fl);

            if (isSuccess) {
                for (Path path : foundFiles) {
                    if (!isLockFile(path) && !tryDelete(path))
                        return false;
                }
            }
        } catch (IOException e) {}
        finally {
            lock.close();
        }

        return isSuccess ? tryDelete(dir) : false;
    }

    /**
     * Register path for cleanup, call cleanup immediately.
     *
     * @param dir           path to clean
     * @param cleanDir      if true, will clean the specified directory, ignoring subdirectories (default).
     * @param subDirRegEx   if not null, will try to clean all subdirectories whose name matches this RegEx (not recursive).
     */
    public static boolean tryCleanup(String dir, boolean cleanDir, String subDirRegEx) {
        return new CleanupPath(Paths.get(dir), cleanDir, subDirRegEx).TryCleanup();
    }

    /**
     * Register path for cleanup, call cleanup immediately.
     *
     * @param dir           path to clean
     */
    public static boolean tryCleanup(String dir) {
        return tryCleanup(dir, true, null);
    }

    /**
     * Activate cleanup procedure for all previously registered paths.
     */
    public static void tryCleanup() {

        synchronized (_cleanupLock) {
            List<CleanupPath> deleted = new ArrayList<>();

            for (CleanupPath p : _cleanupDirs)
                if (p.TryCleanup())
                    deleted.add(p);

            for (CleanupPath p : deleted)
                _cleanupDirs.remove(p);
        }
    }

    /**
     * Register path for cleanup.
     *
     * @param path          path to clean
     * @param cleanDir      if true, will clean the specified directory, ignoring subdirectories (default).
     * @param subDirRegEx   if not null, will try to clean all subdirectories whose name matches this RegEx (not recursive).
     */
    public static void addCleanupPath(Path path, boolean cleanDir, String subDirRegEx) {
        synchronized (_cleanupLock) {
            _cleanupDirs.add(new CleanupPath(path, cleanDir, subDirRegEx));
        }
    }

    /**
     * Register path for cleanup.
     *
     * @param path          path to clean
     */
    public static void addCleanupPath(Path path) {
        addCleanupPath(path, true, null);
    }

    /**
     * Register FileJanitor's on-exit cleanup callback.
     * <p>
     * May be called several times but will only have effect once.
     */
    public static void registerForCleanupOnExit() {

        synchronized (_cleanupLock) {
            if (!_handlerRegistered) {
                Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {
                    public void run() {
                        tryCleanup();
                    }
                }));
                _handlerRegistered = true;
            }
        }
    }

    private static boolean _handlerRegistered;

    private static Object _cleanupLock = new Object();
    private static List<CleanupPath> _cleanupDirs = new ArrayList<>();

    static class CleanupPath {
        private final Path _path;
        private final String _subDirRegEx;
        private final int _flags;

        private final int CLEAN_DIR = 1;
        public boolean TryCleanup() {
            try {
                if (!Files.exists(_path))
                    return true;

                boolean success = true;
                // Clean subdirs?
                if (null != _subDirRegEx) {
                    Pattern p = Pattern.compile(_subDirRegEx);
                    Matcher m = p.matcher("");
                    try (DirectoryStream<Path> paths = Files.newDirectoryStream(_path)) {
                        for (Path childDir : paths) {
                            if (Files.isDirectory(childDir)) {
                                m.reset(childDir.toString());
                                if (m.find())
                                    success &= FileJanitor.tryDeleteDirectory(childDir);
                            }
                        }
                    }
                }

                if (0 != (_flags & CLEAN_DIR))
                    success &= FileJanitor.tryDeleteDirectory(_path);

                return success;
            } catch (IOException e) {
                return false;
            }
        }

        public CleanupPath(Path path, boolean cleanDir, String subDirRegEx) {
            _path = path;
            _subDirRegEx = subDirRegEx;
            _flags = (cleanDir ? CLEAN_DIR : 0);
        }
    }
}


